Denis_Protsenko.py

```diff
@@ -0,0 +1,218 @@
+from pydantic import BaseModel
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Union
+
+
+# Абстрактний клас
+class PublicModel(ABC):
+    @abstractmethod
+    def book_info(self):
+        pass
+
+
+# Клас моделі книги
+class BookModel(BaseModel):
+    title: str
+    author: str
+    year: int
+
+
+# Клас моделі журналу
+class MagazineModel(BaseModel):
+    title: str
+    author: str
+    year: int
+
+
+# Клас книги
+class Book(PublicModel):
+    def __init__(self, book_model: BookModel):
+        self.title = book_model.title
+        self.author = book_model.author
+        self.year = book_model.year
+
+    def book_info(self):  # Implementing the abstract method
+        return f"{self.title}, Автор: {self.author}, Рік видання: {self.year}"
+
+    def __eq__(self, other):
+        if not isinstance(other, Book):
+            return False
+        return (self.title, self.author, self.year) == (other.title, other.author, other.year)
+
+    def __hash__(self):
+        return hash((self.title, self.author, self.year))
+
+
+# Клас журналу
+class Magazine(PublicModel):
```

Comment:
Same here with inheritance, you can define all same functionality in parent class and child classes will reuse it.
Please remind me on next lection to talk about inheritance more.